File: iOS/DemoObjC/ArticleViewController.m
#import "ArticleViewController.h"
#import "DemoModels.h"
#import <MembranaAdWallSDK/MembranaAdWallSDK-Swift.h>

@interface ArticleViewController () <MBRAdWallDelegate>
@property (nonatomic, strong) DemoArticle *article;
@property (nonatomic, strong) MBRAdWall *adWall;
@property (nonatomic, assign) DemoArticleGateScope gateScope;
@property (nonatomic, strong) UIScrollView *scrollView;
@property (nonatomic, strong) UIView *protectedContent;
@property (nonatomic, strong, nullable) MBRAdWallAppearance *appearanceToRestore;
@end

@implementation ArticleViewController

- (instancetype)initWithArticle:(DemoArticle *)article
                          adWall:(MBRAdWall *)adWall
                       gateScope:(DemoArticleGateScope)gateScope {
    self = [super initWithNibName:nil bundle:nil];
    if (self) {
        _article = article;
        _adWall = adWall;
        _gateScope = gateScope;
        switch (gateScope) {
            case DemoArticleGateScopeWholePage: self.title = @"Whole page"; break;
            case DemoArticleGateScopePartialPage: self.title = @"Part of screen"; break;
            case DemoArticleGateScopeBrandedAppearance: self.title = @"Branded styling"; break;
        }
        self.navigationItem.largeTitleDisplayMode = UINavigationItemLargeTitleDisplayModeNever;
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.view.backgroundColor = UIColor.systemBackgroundColor;
    // Publisher integration step 3: add and constrain the content before asking the SDK to gate it.
    [self buildArticle];
    [self applyBrandAppearanceIfNeeded];
    __weak typeof(self) weakSelf = self;
    [self.adWall startWithCompletion:^(NSError *error) {
        __strong typeof(weakSelf) self = weakSelf;
        if (!self) { return; }
        if (error) {
            self.navigationItem.prompt = @"Ads are currently unavailable";
            return;
        }
        UIView *gatedView = self.gateScope == DemoArticleGateScopePartialPage ? self.protectedContent : self.view;
        // Publisher integration step 4: gate the exact protected view, provide its presenting view
        // controller, and implement MBRAdWallDelegate for the terminal access decision.
        [self.adWall gateView:gatedView
                            in:self
                      delegate:self];
    }];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    // Restore the previous style so the other scenarios keep the SDK's own look. A wall already on
    // screen keeps the style it was shown with, so this is safe while a presentation is active.
    if (self.appearanceToRestore != nil) {
        self.adWall.appearance = self.appearanceToRestore;
        self.appearanceToRestore = nil;
    }
}

/// Remote settings always own the wall's text and logo; an appearance only restyles how that
/// content is drawn.
///
/// The per-presentation `appearance:` overload of `gate` is Swift-only, so Objective-C callers set
/// the instance-level property instead. It applies to every wall this instance presents from now
/// on, which is why this screen restores the previous value on the way out. Set it before
/// `startWithCompletion:` rather than inside the completion, so navigating away early cannot leave
/// the override in place.
- (void)applyBrandAppearanceIfNeeded {
    if (self.gateScope != DemoArticleGateScopeBrandedAppearance) { return; }
    self.appearanceToRestore = self.adWall.appearance;
    // Objective-C has no copy-style initializer, so every value is listed. Everything except the
    // call to action repeats the SDK's own defaults.
    self.adWall.appearance = [[MBRAdWallAppearance alloc]
        initWithBackdropColor:UIColor.systemBackgroundColor
          cardBackgroundColor:UIColor.secondarySystemBackgroundColor
                   titleColor:UIColor.labelColor
             descriptionColor:UIColor.secondaryLabelColor
        buttonBackgroundColor:UIColor.systemPinkColor
        buttonForegroundColor:UIColor.whiteColor
             closeButtonColor:UIColor.labelColor
                    titleFont:[UIFont fontWithName:@"AvenirNext-Bold" size:22]
              descriptionFont:[UIFont fontWithName:@"AvenirNext-Regular" size:17]
                   buttonFont:[UIFont fontWithName:@"AvenirNext-DemiBold" size:17]];
}

- (void)buildArticle {
    self.scrollView = [[UIScrollView alloc] init];
    self.scrollView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:self.scrollView];
    [NSLayoutConstraint activateConstraints:@[
        [self.scrollView.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor],
        [self.scrollView.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor],
        [self.scrollView.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor],
        [self.scrollView.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor]
    ]];

    UIStackView *stack = [[UIStackView alloc] init];
    stack.translatesAutoresizingMaskIntoConstraints = NO;
    stack.axis = UILayoutConstraintAxisVertical;
    stack.spacing = 18;
    stack.layoutMargins = UIEdgeInsetsMake(20, 24, 48, 24);
    stack.layoutMarginsRelativeArrangement = YES;
    [self.scrollView addSubview:stack];
    [NSLayoutConstraint activateConstraints:@[
        [stack.leadingAnchor constraintEqualToAnchor:self.scrollView.contentLayoutGuide.leadingAnchor],
        [stack.trailingAnchor constraintEqualToAnchor:self.scrollView.contentLayoutGuide.trailingAnchor],
        [stack.topAnchor constraintEqualToAnchor:self.scrollView.contentLayoutGuide.topAnchor],
        [stack.bottomAnchor constraintEqualToAnchor:self.scrollView.contentLayoutGuide.bottomAnchor],
        [stack.widthAnchor constraintEqualToAnchor:self.scrollView.frameLayoutGuide.widthAnchor]
    ]];

    UILabel *category = [self label:self.article.category style:UIFontTextStyleCaption1 color:self.article.color];
    UILabel *title = [self label:self.article.title style:UIFontTextStyleTitle1 color:UIColor.labelColor];
    UILabel *summary = [self label:self.article.summary style:UIFontTextStyleTitle3 color:UIColor.secondaryLabelColor];
    [stack addArrangedSubview:category];
    [stack addArrangedSubview:title];
    [stack addArrangedSubview:summary];

    UIView *hero = [[UIView alloc] init];
    hero.backgroundColor = self.article.color;
    hero.layer.cornerRadius = 12;
    hero.clipsToBounds = YES;
    UIImageView *symbol = [[UIImageView alloc] initWithImage:[UIImage systemImageNamed:self.article.symbolName]];
    symbol.translatesAutoresizingMaskIntoConstraints = NO;
    symbol.tintColor = UIColor.whiteColor;
    symbol.contentMode = UIViewContentModeScaleAspectFit;
    [hero addSubview:symbol];
    [NSLayoutConstraint activateConstraints:@[
        [hero.heightAnchor constraintEqualToConstant:180],
        [symbol.centerXAnchor constraintEqualToAnchor:hero.centerXAnchor],
        [symbol.centerYAnchor constraintEqualToAnchor:hero.centerYAnchor],
        [symbol.widthAnchor constraintEqualToConstant:72],
        [symbol.heightAnchor constraintEqualToConstant:72]
    ]];
    if (self.article.paragraphs.count > 0) {
        [stack addArrangedSubview:[self label:self.article.paragraphs.firstObject
                                        style:UIFontTextStyleBody
                                        color:UIColor.labelColor]];
    }

    self.protectedContent = [[UIView alloc] init];
    UIStackView *protectedStack = [[UIStackView alloc] init];
    protectedStack.translatesAutoresizingMaskIntoConstraints = NO;
    protectedStack.axis = UILayoutConstraintAxisVertical;
    protectedStack.spacing = 18;
    [protectedStack addArrangedSubview:hero];
    for (NSUInteger index = 1; index < self.article.paragraphs.count; index++) {
        [protectedStack addArrangedSubview:[self label:self.article.paragraphs[index] style:UIFontTextStyleBody color:UIColor.labelColor]];
    }
    [self.protectedContent addSubview:protectedStack];
    [NSLayoutConstraint activateConstraints:@[
        [protectedStack.leadingAnchor constraintEqualToAnchor:self.protectedContent.leadingAnchor],
        [protectedStack.trailingAnchor constraintEqualToAnchor:self.protectedContent.trailingAnchor],
        [protectedStack.topAnchor constraintEqualToAnchor:self.protectedContent.topAnchor],
        [protectedStack.bottomAnchor constraintEqualToAnchor:self.protectedContent.bottomAnchor]
    ]];
    [stack addArrangedSubview:self.protectedContent];
}

- (UILabel *)label:(NSString *)text style:(UIFontTextStyle)style color:(UIColor *)color {
    UILabel *label = [[UILabel alloc] init];
    label.text = text;
    label.font = [UIFont preferredFontForTextStyle:style];
    label.adjustsFontForContentSizeCategory = YES;
    label.textColor = color;
    label.numberOfLines = 0;
    return label;
}

- (void)adWallDidUnlockContent:(MBRAdWall *)adWall {
    // Publisher integration step 5: this callback is the successful access grant.
    // Deferred one run loop turn: see the comment in ScenarioListViewController.
    dispatch_async(dispatch_get_main_queue(), ^{
        self.navigationItem.prompt = @"Unlocked by a rewarded ad";
    });
}

- (void)adWallDidRejectContent:(MBRAdWall *)adWall {
    // Keep content protected or install the publisher's fallback when access is not earned.
    dispatch_async(dispatch_get_main_queue(), ^{
        self.navigationItem.prompt = @"Publisher handled rejection";
    });
}

- (void)adWall:(MBRAdWall *)adWall contentUnavailableWithReason:(MBRUnavailableReason)reason {
    // Capping and no-fill are expected production states. Keep premium content protected and
    // provide the publisher's own login, subscription, or retry path.
    dispatch_async(dispatch_get_main_queue(), ^{
        self.navigationItem.prompt = @"No wall available; publisher fallback applied";
    });
}

- (void)adWall:(MBRAdWall *)adWall didFailWithError:(NSError *)error {
    // Send technical failures to publisher diagnostics, then keep content protected or replace it
    // with an app-owned fallback. Never unlock from this callback.
    dispatch_async(dispatch_get_main_queue(), ^{
        self.navigationItem.prompt = @"Ad error; publisher fallback applied";
    });
}

@end

File: iOS/DemoUIKit/ArticleViewController.swift
import MembranaAdWallSDK
import UIKit

enum ArticleGateScope {
    case wholePage
    case partialPage
    /// Same whole-page coverage as `wholePage`; only the wall's styling differs.
    case brandedAppearance

    var navigationTitle: String {
        switch self {
        case .wholePage: return "Whole page"
        case .partialPage: return "Part of screen"
        case .brandedAppearance: return "Branded styling"
        }
    }

    var coversWholePage: Bool {
        self != .partialPage
    }

    /// Remote settings always own the wall's text and logo. An appearance only restyles how that
    /// content is drawn, so a publisher can match the card to its own design language. Returning
    /// `nil` leaves the SDK's own look in place.
    var appearance: MembranaAdWallAppearance? {
        guard self == .brandedAppearance else { return nil }
        return MembranaAdWallAppearance(
            buttonBackgroundColor: .systemPink,
            buttonForegroundColor: .white,
            titleFont: UIFont(name: "AvenirNext-Bold", size: 22),
            descriptionFont: UIFont(name: "AvenirNext-Regular", size: 17),
            buttonFont: UIFont(name: "AvenirNext-DemiBold", size: 17)
        )
    }
}

final class ArticleViewController: UIViewController, MembranaAdWallDelegate {
    private let article: DemoArticle
    private let adWall: MembranaAdWall
    private let gateScope: ArticleGateScope
    private let scrollView = UIScrollView()
    private let stack = UIStackView()
    private let protectedContent = UIView()
    private var gateTask: Task<Void, Never>?

    init(article: DemoArticle, adWall: MembranaAdWall, gateScope: ArticleGateScope = .partialPage) {
        self.article = article
        self.adWall = adWall
        self.gateScope = gateScope
        super.init(nibName: nil, bundle: nil)
        title = gateScope.navigationTitle
        navigationItem.largeTitleDisplayMode = .never
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) { fatalError("init(coder:) has not been implemented") }

    override func viewDidLoad() {
        super.viewDidLoad()
        view.backgroundColor = .systemBackground
        // Publisher integration step 3: add and fully constrain the content that must remain
        // protected before asking the SDK to gate it.
        configureArticle()
        gateWhenAdWallIsReady()
    }

    deinit {
        gateTask?.cancel()
    }

    private func gateWhenAdWallIsReady() {
        let adWall = adWall
        gateTask = Task { @MainActor [weak self] in
            do {
                try await adWall.start()
                guard !Task.isCancelled, let self else { return }
                // Publisher integration step 4: gate either the complete screen or the exact
                // protected subtree, and provide the presenting view controller plus delegate.
                adWall.gate(
                    gateScope.coversWholePage ? view : protectedContent,
                    in: self,
                    appearance: gateScope.appearance,
                    delegate: self
                )
            } catch {
                guard let self else { return }
                demoSetPrompt("Ads are currently unavailable")
            }
        }
    }

    private func configureArticle() {
        scrollView.translatesAutoresizingMaskIntoConstraints = false
        scrollView.alwaysBounceVertical = true
        view.addSubview(scrollView)
        NSLayoutConstraint.activate([
            scrollView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
            scrollView.trailingAnchor.constraint(equalTo: view.trailingAnchor),
            scrollView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor),
            scrollView.bottomAnchor.constraint(equalTo: view.bottomAnchor)
        ])

        stack.translatesAutoresizingMaskIntoConstraints = false
        stack.axis = .vertical
        stack.spacing = 18
        stack.isLayoutMarginsRelativeArrangement = true
        stack.layoutMargins = UIEdgeInsets(top: 20, left: 24, bottom: 48, right: 24)
        scrollView.addSubview(stack)
        NSLayoutConstraint.activate([
            stack.leadingAnchor.constraint(equalTo: scrollView.contentLayoutGuide.leadingAnchor),
            stack.trailingAnchor.constraint(equalTo: scrollView.contentLayoutGuide.trailingAnchor),
            stack.topAnchor.constraint(equalTo: scrollView.contentLayoutGuide.topAnchor),
            stack.bottomAnchor.constraint(equalTo: scrollView.contentLayoutGuide.bottomAnchor),
            stack.widthAnchor.constraint(equalTo: scrollView.frameLayoutGuide.widthAnchor)
        ])

        let category = makeLabel(article.category, style: .caption1, color: article.color)
        let title = makeLabel(article.title, style: .title1)
        let summary = makeLabel(article.summary, style: .title3, color: .secondaryLabel)
        let hero = UIView()
        hero.backgroundColor = article.color
        hero.layer.cornerRadius = 12
        hero.clipsToBounds = true
        let symbol = UIImageView(image: UIImage(systemName: article.symbolName))
        symbol.translatesAutoresizingMaskIntoConstraints = false
        symbol.tintColor = .white
        symbol.contentMode = .scaleAspectFit
        hero.addSubview(symbol)
        NSLayoutConstraint.activate([
            hero.heightAnchor.constraint(equalToConstant: 180),
            symbol.centerXAnchor.constraint(equalTo: hero.centerXAnchor),
            symbol.centerYAnchor.constraint(equalTo: hero.centerYAnchor),
            symbol.widthAnchor.constraint(equalToConstant: 72),
            symbol.heightAnchor.constraint(equalToConstant: 72)
        ])

        [category, title, summary].forEach(stack.addArrangedSubview)

        if let introduction = article.paragraphs.first {
            stack.addArrangedSubview(makeLabel(introduction, style: .body))
        }

        let protectedStack = UIStackView()
        protectedStack.translatesAutoresizingMaskIntoConstraints = false
        protectedStack.axis = .vertical
        protectedStack.spacing = 18
        protectedStack.addArrangedSubview(hero)
        article.paragraphs.dropFirst().forEach {
            protectedStack.addArrangedSubview(makeLabel($0, style: .body))
        }
        protectedContent.addSubview(protectedStack)
        NSLayoutConstraint.activate([
            protectedStack.leadingAnchor.constraint(equalTo: protectedContent.leadingAnchor),
            protectedStack.trailingAnchor.constraint(equalTo: protectedContent.trailingAnchor),
            protectedStack.topAnchor.constraint(equalTo: protectedContent.topAnchor),
            protectedStack.bottomAnchor.constraint(equalTo: protectedContent.bottomAnchor)
        ])
        stack.addArrangedSubview(protectedContent)
    }

    private func makeLabel(_ text: String, style: UIFont.TextStyle, color: UIColor = .label) -> UILabel {
        let label = UILabel()
        label.text = text
        label.font = .preferredFont(forTextStyle: style)
        label.adjustsFontForContentSizeCategory = true
        label.textColor = color
        label.numberOfLines = 0
        return label
    }

    func adWallDidUnlockContent(_ adWall: MembranaAdWall) {
        // Publisher integration step 5: treat this callback as the successful access grant.
        demoSetPrompt("Unlocked by a rewarded ad")
    }

    func adWallDidRejectContent(_ adWall: MembranaAdWall) {
        // Keep content protected or install the publisher's fallback when access is not earned.
        demoSetPrompt("Ad declined; publisher fallback applied")
    }

    func adWall(
        _ adWall: MembranaAdWall,
        contentUnavailableWithReason reason: MembranaAdWallUnavailableReason
    ) {
        // Routine states such as capping and no-fill belong here. Keep premium content protected
        // and offer the publisher's own login, subscription, or retry path.
        demoSetPrompt("No wall available; publisher fallback applied")
    }

    func adWall(_ adWall: MembranaAdWall, didFailWithError error: NSError) {
        // Report the technical error through the publisher's diagnostics, then keep content
        // protected or replace it with an app-owned fallback.
        demoSetPrompt("Ad error; publisher fallback applied")
    }
}
